sparse_mask: backward support for sparse lhs (take 2)

nikitaved opened the pull request:
This is a copy of https://github.com/pytorch/pytorch/pull/95165 with some bug fixes.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #99298
* #99043
* #98490
* __->__ #104341



cc @alexsamardzic @pearu @cpuhrsch @amjames @bhosmer @ezyang @albanD @zou3519 @gqchen @soulitzer @Lezcano @Varal7

nikitaved committed:
sparse_mask: backward support for sparse lhs (take 2)

[ghstack-poisoned]

nikitaved commented:
Sorry to bother you again with this one, @pearu, but could you please approve this one again?

nikitaved commented:
@albanD , @soulitzer , could you please also approve it for good measure since @cpuhrsch is on PTO?

nikitaved committed:
Update on "sparse_mask: backward support for sparse lhs (take 2)"

This is a copy of https://github.com/pytorch/pytorch/pull/95165 with some bug fixes.




cc alexsamardzic pearu cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

nikitaved committed:
Update on "sparse_mask: backward support for sparse lhs (take 2)"

This is a copy of https://github.com/pytorch/pytorch/pull/95165 with some bug fixes.




cc alexsamardzic pearu cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

nikitaved committed:
Update on "sparse_mask: backward support for sparse lhs (take 2)"

This is a copy of https://github.com/pytorch/pytorch/pull/95165 with some bug fixes.




cc alexsamardzic pearu cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

albanD reviewed (approved):
Looks good!

amjames commented on test/test_sparse.py:
```diff
@@ -2140,6 +2141,36 @@ def _test_sparse_mask_hybrid_fixed():
+                    lhs_data.shape
+                )._coalesced_(rhs_is_coalesced)
+
+                # setting masked = True is required because of the broken backward of to_dense().
```
Does this work as expected without using the `masked=True` kwarg for gradcheck, if you instead use the `masked_grad=False` kwarg to `to_dense` ?

nikitaved replied:
Yes, this change should be there up in the stack. But let me add it here for completeness.

nikitaved committed:
Update on "sparse_mask: backward support for sparse lhs (take 2)"

This is a copy of https://github.com/pytorch/pytorch/pull/95165 with some bug fixes.




cc alexsamardzic pearu cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

nikitaved commented on test/test_sparse.py:
```diff
@@ -2140,6 +2141,36 @@ def _test_sparse_mask_hybrid_fixed():
+
+                # setting masked = True is required because of the broken backward of to_dense().
+                # See https://github.com/pytorch/pytorch/issues/95550.
+                gradcheck(lambda x: x.sparse_mask(rhs).to_dense(), (lhs,), masked=True, check_sparse_nnz=True)
```
Removed.

pearu reviewed (approved):
LGTM! Thanks, @nikitaved!

amjames reviewed (approved):
Thanks!

nikitaved committed:
Update on "sparse_mask: backward support for sparse lhs (take 2)"

This is a copy of https://github.com/pytorch/pytorch/pull/95165 with some bug fixes.




cc alexsamardzic pearu cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

nikitaved commented:
Alright, it looks weird. Slowtests are failing in somewhat non-deterministic order across the shards. Is there a way to trigger these things locally? I have no issues when running gradcheck with `fast_mode=False`.

nikitaved committed:
Update on "sparse_mask: backward support for sparse lhs (take 2)"

This is a copy of https://github.com/pytorch/pytorch/pull/95165 with some bug fixes.




cc alexsamardzic pearu cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

pearu commented:
> Alright, it looks weird. Slowtests are failing in somewhat non-deterministic manner across shards. Is there a way to trigger these things locally? I have no issues when running gradcheck with `fast_mode=False`.

Yes, try defining env variable `PYTORCH_TEST_WITH_SLOW_GRADCHECK=1` prior running the gradcheck tests.

nikitaved commented:
Thank you, @pearu! Now it fails locally, but only when `masked=True`...

nikitaved committed:
Update on "sparse_mask: backward support for sparse lhs (take 2)"

This is a copy of https://github.com/pytorch/pytorch/pull/95165 with some bug fixes.




cc alexsamardzic pearu cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

nikitaved commented:
Nice, problem solved! One more for `masked=True` being very confusing ;)

nikitaved committed:
Update on "sparse_mask: backward support for sparse lhs (take 2)"

This is a copy of https://github.com/pytorch/pytorch/pull/95165 with some bug fixes.




cc alexsamardzic pearu cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

nikitaved committed:
Update on "sparse_mask: backward support for sparse lhs (take 2)"

This is a copy of https://github.com/pytorch/pytorch/pull/95165 with some bug fixes.




cc alexsamardzic pearu cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

nikitaved committed:
Update on "sparse_mask: backward support for sparse lhs (take 2)"

This is a copy of https://github.com/pytorch/pytorch/pull/95165 with some bug fixes.




cc alexsamardzic pearu cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

nikitaved committed:
Update on "sparse_mask: backward support for sparse lhs (take 2)"

This is a copy of https://github.com/pytorch/pytorch/pull/95165 with some bug fixes.




cc alexsamardzic pearu cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

nikitaved commented:
The failures are unrelated :) I think it is good to go!

amjames commented:
@pytorchbot  merge

nikitaved commented:
@amjames , I will rebase, then we can merge :)

nikitaved committed:
Update on "sparse_mask: backward support for sparse lhs (take 2)"

This is a copy of https://github.com/pytorch/pytorch/pull/95165 with some bug fixes.




cc alexsamardzic pearu cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

nikitaved committed:
Update on "sparse_mask: backward support for sparse lhs (take 2)"

This is a copy of https://github.com/pytorch/pytorch/pull/95165 with some bug fixes.




cc alexsamardzic pearu cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]
